biz.aQute.bndlib/src/aQute/bnd/cdi/AnnotationReader.java

```
@@ -0,0 +1,340 @@
+		Instruction marked;
+
+		@Override
+		public void annotation(Annotation annotation) throws Exception {
+			java.lang.annotation.Annotation a = annotation.getAnnotation();
+
+			if (a instanceof Beans) {
+				Object[] beanClasses = annotation.get("value");
+
+				if (beanClasses != null && beanClasses.length > 0) {
+					marked = new Instruction(Arrays.stream(beanClasses)
+						.map(c -> c.toString())
+						.collect(Collectors.joining(",")));
+				} else {
+					marked = new Instruction("*");
+				}
+			}
+		}
+	}
+
+	@Override
+	public void annotation(Annotation annotation) {
+		try {
+			java.lang.annotation.Annotation a = annotation.getAnnotation();
+
+			if (a instanceof Bean) {
+				definitions.get(0).marked = true;
+			}
+			else if (a instanceof Service) {
+				doService(annotation, a);
+			}
+			else if (a instanceof MinimumCardinality) {
+				minimumCardinality = (int) annotation.get("value");
+				if (minimumCardinality > 0 && referenceDef != null) {
+					referenceDef.cardinality = ReferenceCardinality.AT_LEAST_ONE;
+				}
+			}
+			if (a instanceof Reference) {
+				doReference(annotation, a);
+			}
+		} catch (Exception e) {
+			analyzer.exception(e, "During bean processing on class %s, exception %s", clazz, e);
+		}
+	}
+
+	private void doReference(Annotation annotation, java.lang.annotation.Annotation a) {
+		referenceDef = new ReferenceDef(annotation, member, analyzer);
+		if (minimumCardinality > 0) {
+			referenceDef.cardinality = ReferenceCardinality.AT_LEAST_ONE;
+			minimumCardinality = -1;
+		}
+		definitions.get(0).references.add(referenceDef);
+	}
+
+	private void doService(Annotation annotation, java.lang.annotation.Annotation a) {
+		switch (annotation.getElementType()) {
+			case FIELD :
+				if (member instanceof Clazz.FieldDef) {
+					String signature = member.getSignature();
+					if (signature == null) {
+						signature = member.getDescriptor()
+							.toString();
+					}
+					FieldSignature sig = FieldSignature.of(signature);
```

This is wrong. You should call `analyzer.getFieldSignature(signature)` to used cached results. Also not that using the method descriptor for a field signature can result in an exception trying to create the signature of the field of of a base type. See 

https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L941-L953

If the field signature cannot be created, the field is a base type and you should probably report an error in the analyzer. For example: https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L1075-L1080